src/lib/SaferNames/Inference.hs

```diff
@@ -1171,22 +1268,56 @@ bindLamPat (WithSrcB pos pat) v cont = addSrcContext pos $ case pat of
     xs <- forM idxs \i -> emit $ App (Var v) i
     bindLamPats ps xs cont
 
-checkAnn :: (EmitsBoth o, Inferer m) => Maybe (UType i) -> m i o (Type o)
+checkAnn :: (EmitsInf o, Inferer m) => Maybe (UType i) -> m i o (Type o)
 checkAnn ann = case ann of
   Just ty -> checkUType ty
   Nothing -> freshType TyKind
 
-checkUType :: EmitsInf o => Inferer m => UType i -> m i o (Type o)
-checkUType ty@(WithSrcE pos _) = addSrcContext pos $ do
-  Abs decls result <- buildDeclsInf $ withAllowedEffects Pure $ checkRho ty TyKind
-  cheapReduceWithDecls decls result >>= \case
-    (_       , Just (d:_)) ->
-      throw TypeErr $ "Couldn't synthesize a class dictionary for: " ++ pprint d
-    (_       , Nothing   ) ->
-      throw TypeErr $ "Couldn't synthesize a class dictionary required by: " ++ pprint ty
-    (Nothing , Just []   ) ->
-      throw TypeErr $ "Can't reduce type expression: " ++ pprint ty
-    (Just ty', Just []   ) -> return ty'
+checkAnnWithMissingDicts :: (EmitsInf o, Inferer m)
+                         => Maybe (UType i)
+                         -> ([Type o] -> (forall o'. (EmitsInf o', Ext o o') => m i o' (Type o')) -> m i o a)
```

Comment:
It takes a bit of work to understand these rank-3 types. Maybe at least describe what the `forall o'. (EmitsInf o', Ext o o') => m i o' (Type o')` argument to the continuation is supposed to be. It seems to be either "generate fresh type" or "check type, reducing again"? If it was me, I'd probably just defunctionalize by making an ADT for those two cases.

An alias like `type DictType = Type` might also help understand some of these signatures.

Reply:
Yeah it took some effort to write out too, but in the end I don't think it's too bad. The type alias is a very good idea and I'll try to add it in a few places like this one!